File: strategies/tldr.go
// Copyright 2025 Naren Yellavula
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package strategies

import (
	"fmt"
	"io"
	"net/http"
)

// TldrStrategy fetches help from TLDR pages - prioritized for cleaner examples
type TldrStrategy struct{}

func (t *TldrStrategy) SupportsCommand(baseCmd string) bool {
	return true // Supports any command as it's a universal fallback
}

func (t *TldrStrategy) Priority() int {
	return 0 // Highest priority - try first for better user experience
}

func (t *TldrStrategy) GetHelp(cmdParts []string) (string, error) {
	cmd := NewCommand(cmdParts)

	baseUrl := "https://raw.githubusercontent.com/tldr-pages/tldr/refs/heads/main/pages/common"
	var fullURL string

	// Support up to 2 levels of sub-commands for TLDR
	if cmd.HasSubCommand(1) {
		subCmd := cmd.GetSubCommand(0)
		fullURL = fmt.Sprintf("%s/%s-%s.md", baseUrl, cmd.BaseCmd, subCmd)
	} else {
		fullURL = fmt.Sprintf("%s/%s.md", baseUrl, cmd.BaseCmd)
	}

	client := &http.Client{Timeout: HttpTimeout}
	resp, err := client.Get(fullURL)
	if err != nil {
		return "", fmt.Errorf("failed to fetch TLDR page: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("TLDR page not found (HTTP %d)", resp.StatusCode)
	}

	limitedReader := io.LimitReader(resp.Body, MaxTldrSize)
	body, err := io.ReadAll(limitedReader)
	if err != nil {
		return "", fmt.Errorf("failed to read TLDR response: %v", err)
	}

	content := string(body)
	if content != "" {
		content = "📚 TLDR Documentation:\n\n" + content
	}

	return content, nil
}

File: recaller-rs/src/help_system/strategies/tldr.rs
use std::io::Read;
use std::time::Duration;

use anyhow::{Context, Result, anyhow};

use crate::help_system::strategy::{CommandParts, HelpStrategy};

const TLDR_TIMEOUT: Duration = Duration::from_secs(10);
const MAX_TLDR_SIZE: usize = 512 * 1024;

pub struct TldrStrategy;

impl TldrStrategy {
    pub fn new() -> Self {
        Self
    }
}

impl HelpStrategy for TldrStrategy {
    fn priority(&self) -> i32 {
        0
    }

    fn supports_command(&self, _: &str) -> bool {
        true
    }

    fn get_help(&self, command: &CommandParts) -> Result<String> {
        let base = command
            .base_cmd()
            .ok_or_else(|| anyhow!("no base command provided"))?;

        let base_url =
            "https://raw.githubusercontent.com/tldr-pages/tldr/refs/heads/main/pages/common";
        let path = if command.has_sub_command(1) {
            format!(
                "{}/{}-{}.md",
                base_url,
                base,
                command.get_sub_command(0).unwrap()
            )
        } else {
            format!("{}/{}.md", base_url, base)
        };

        let response = ureq::get(&path)
            .timeout(TLDR_TIMEOUT)
            .call()
            .with_context(|| format!("failed to fetch TLDR page from {path}"))?;

        if response.status() != 200 {
            return Err(anyhow!(
                "TLDR page not found for {base} (HTTP {})",
                response.status()
            ));
        }

        let mut reader = response.into_reader();
        let mut buffer = Vec::new();
        let mut chunk = [0u8; 8192];
        while buffer.len() < MAX_TLDR_SIZE {
            let read_len = std::cmp::min(chunk.len(), MAX_TLDR_SIZE - buffer.len());
            match reader.read(&mut chunk[..read_len]) {
                Ok(0) => break,
                Ok(n) => buffer.extend_from_slice(&chunk[..n]),
                Err(err) => return Err(anyhow!("failed to read TLDR response: {err}")),
            }
        }

        let mut content = String::from_utf8_lossy(&buffer).to_string();
        if content.is_empty() {
            return Err(anyhow!("TLDR page empty"));
        }
        content = format!("📚 TLDR Documentation:\n\n{}", content);
        Ok(content)
    }
}
